### ISSUE TITLE
ENH: integrate: evaluate simultaneously a function and its jacobian for ODE integration

### ISSUE THREAD

**Issue by DrizztSama** [scipy/scipy#20903]
### Is your feature request related to a problem? Please describe.

When doing numerical approximation of evolution PDE of the form $\partial_t u=Du$ where $D$ is a spatial differential operator we often use a "tensorial" approach by using a numerical method to discretize first the space variable and the spatial operator (using FE, FD, FV…) which leads to a usually high dimensional ODE of the form $U'=F(t, U)$ on which you can now apply the standard tools of ODE approximation, "forgetting" the original PDE. However, what you *don't* lose in the process is that the evaluation of $F(t, U)$ can be a bit costly. Even in the simplest cases where $F$ is linear with respect to the state variable one can obtain $F(t, U)=A(t)U+b(t)$ where builiding the matrice $A(t)$ and vector $b(t)$ can be an endeavour.

TL;DR When I try to solve $y'=A(t)y+b(t)$ with implicit method I have to separate the evaluation of the function $A(t)y+b(t)$ and the the evaluation of the jacobian $A(t)$ when both require to compute the matrix $A(t)$ which, depending on the problem, can be slow/costly.

Note that this only an issue here when $A$ actually depends on time: if not, one can already sort this problem out by using the `args` kwarg in `solve_ivp` (or directly to the numerical solver).

### Describe the solution you'd like.

It seems to me (but I did not review all the implemented methods) that numerical methods that rely on the use of the jacobian of the function $F(t, y)$ would ever evaluate the jacobian $dF(t, y)$ at the same arguments as $F$. So the simplest way to account for this issue (from a user standpoint!) would be to  simply allow the user to give the solver an object that computes both $F(t, y)$ and $dF(t, y)$ simultaneously for the same arguments $(t, y)$. 

To not modify the current signature of ODE integrators I would be happy with an added optional argument to pass a callable that does just that.

However I am very aware that this would need quite the modification in the codebase to account for this supplementary option.

### Describe alternatives you've considered.

I am not a programmer per say so there might be fixes (obvious or subtle) that I am not aware of: like maybe there are ways to code $F$ and $dF$ that can factor the common work while keeping the signature needed for the Scipy ODE solvers.

To illustrate on that, I tried something along the line of $F(t, y A)=A(t)y+b(t)$ and $dF(t, y, A)=A(t)$ and pass via `args` the function computing $t\mapsto A(t)$. This, I believe, does not avoid the double computing of $A(t)$ but there may be a smarter way to do something alike that I am not aware of.

**Comment by mdhaber** [comment]
This was discussed recently in gh-19305 in the context of optimization, but I think it would be worthwhile to review the discussion. The crux of it is that several methods evaluate the function and its derivative(s) different numbers of times, so if you combine the function and jacobian evaluation into one callable, you might end up being less efficient than you could be. For instance:
```python3
from scipy import integrate

def f(t, z):
    f.nit += 1
    x, y = z
    a, b, c, d = (1.5, 1, 3, 1)
    return [a*x - b*x*y, -c*y + d*x*y]
f.nit = 0

def df(t, z):
    df.nit += 1
    x, y = z
    a, b, c, d = (1.5, 1, 3, 1)
    return [[a - b*y, -b*x], [d*y, -c + d*x]]
df.nit = 0

sol = integrate.solve_ivp(f, [0, 15], [10, 5], method='BDF')  # no jacobian passed
f.nit  # 669
```
If we re-run passing the jacobian, the function is still called 639 times, and the jacobian is called only ten times. So if you combine the callables, you would be computing the jacobian hundreds of times more than necessary. The solution is to write a class that maintains its state and have function and jacobian methods that memoize results.

**Comment by DrizztSama** [comment]
> IIUC, this was discussed recently in [gh-19305](https://github.com/scipy/scipy/issues/19305) in the context of optimization, but I think it would be worthwhile to review that discussion. The crux of it is that several methods evaluate the function and its derivative(s) different numbers of times, so if you combine the function and jacobian evaluation into one callable, you might end up being _less_ efficient than you would have bee. For instance, adapting the Lotka-Volterra example from the documentation:
> 
> ```python
> from scipy import integrate
> 
> def f(t, z):
>     f.nit += 1
>     x, y = z
>     a, b, c, d = (1.5, 1, 3, 1)
>     return [a*x - b*x*y, -c*y + d*x*y]
> f.nit = 0
> 
> def df(t, z):
>     df.nit += 1
>     x, y = z
>     a, b, c, d = (1.5, 1, 3, 1)
>     return [[a - b*y, -b*x], [d*y, -c + d*x]]
> df.nit = 0
> 
> sol = integrate.solve_ivp(f, [0, 15], [10, 5], method='BDF')  # no jacobian passed
> f.nit  # 669
> ```
> 
> If we re-run with `jac=df`, the function is called 639 times, and the jacobian is called only ten times. So if you combine the callables, you would end up performing the additional jacobian part of the calculation hundreds of times more than necessary, whereas now you just perform the function part of the calculation ten times more than necessary. You get the best of both worlds by writing a class with function and jacobian methods that memoize partial results.

I definitely did not think of that. Thanks a lot for pointing me in this direction @mdhaber. From my perspective, this discussion can thus be closed.

However since the topic seem to come up from time to time (albeit in different areas of scientific computations) it may be worth writing a little something on the matter somewhere in the documentation.

Sorry for the inconvenience.
